Refactor of chapter 6 and edits to chapter 5

jrgold opened the pull request:
Changes to chapter 6:
Removed most introductory pointer material since it is covered
    in chapter 5
Changed the structure and ordering of the chapter to teach all
    concepts required for dynamic arrays (array/pointer mingling,
    array memory layout, malloc) before seeing dynamic arrays.
    I was concerned with jumping straight into malloc as a solution
    before covering the "why" and "how".
Made regular edits (clarifications, wording)
Added a section on initializing arrays
Changed "2dNums" variable name to one that compiles
Added a clarification about row/column major layout not being fixed
Covered an array's memory layout more explicitly
Moved and reworked the pointer arithmetic section and tried to relate
    it more towards arrays in general than malloc
Introduced malloc as just an allocation mechanism, before tying it
    in with dynamic arrays to prevent it from looking like a magic bullet
Indtroduced sizeof more explicitly
Mentioned the second way of dynamically allocating 2d arrays

Changes to chapter 5:
Changed the title since "Passing Types" was a bit confusing
Minor formatting changes
Added an example of a pointer-to-pointer since I removed it from the
    arrays chapter

Punctuation fix in Operators.

jrgold committed:
Refactor of chapter 6 and edits to chapter 5

Changes to chapter 6:
Removed most introductory pointer material since it is covered
	in chapter 5
Changed the structure and ordering of the chapter to teach all
	concepts required for dynamic arrays (array/pointer mingling,
	array memory layout, malloc) before seeing dynamic arrays.
	I was concerned with jumping straight into malloc as a solution
	before covering the "why" and "how".
Made regular edits (clarifications, wording)
Added a section on initializing arrays
Changed "2dNums" variable name to one that compiles
Added a clarification about row/column major layout not being fixed
Covered an array's memory layout more explicitly
Moved and reworked the pointer arithmetic section and tried to relate
	it more towards arrays in general than malloc
Introduced malloc as just an allocation mechanism, before tying it
	in with dynamic arrays to prevent it from looking like a magic bullet
Indtroduced sizeof more explicitly
Mentioned the second way of dynamically allocating 2d arrays

Changes to chapter 5:
Changed the title since "Passing Types" was a bit confusing
Minor formatting changes
Added an example of a pointer-to-pointer since I removed it from the
	arrays chapter

Punctuation fix in Operators.

jrgold commented:
This is a pretty major change, and I'd like to get feedback on it. It's a pretty big restructure with some large removals and additions, so I'm happy to change anything.

The "Close pull request" button should not be write next to the "Comment" button where cancel would normally go.

mon commented on arraysandpointers.tex:
```diff
@@ -1,144 +1,164 @@
-int* p1 = p; // p1 points to the same thing p does.
+/* These are both the same */
+*(array + 1) = 1;
+array[1] = 1; 
```
Perhaps mention that square brackets are the more common and preferred method for this course?

mon commented on arraysandpointers.tex:
```diff
@@ -210,63 +191,67 @@ \subsection{Multidimensional Dynamic Array}
+
+As with the nested arrays we saw earlier in the chapter, there is nothing special about our choice of \lstinline!arr! as an array of rows rather than an array of columns.
+
+The other commonly used method for dynamically-sized multidimensional arrays is to allocate a single block of memory large enough to hold every element, and calculate the indexes required for specific elements.
```
Perhaps an explanation on why you'd use this over the conceptually easier method

mon commented on functions.tex:
```diff
@@ -101,7 +103,8 @@ \chapter{Functions and Passing types}
+int * q; // Now you're just being silly. (Still legal though)
+int** m; // Pointers to pointers are okay too
 \end{codeblock}
 The last two lines are examples of legal code which is unnecessarily hard to read.
```
No longer last 2 since you added the *\* example, probably just move m above the badly spaced ones

mon commented on functions.tex:
```diff
@@ -117,6 +120,9 @@ \chapter{Functions and Passing types}
 Most importantly, you need to conform with whatever style rules are in force for your organisation.
 
+Line~7 is an example of a pointer to a pointer.
+Since a pointers is a regular value, it must reside in memory, and so can be referred to with another pointer.
```
s/pointers/pointer

jrgold committed:
Mon's edits and clarifications to ch. 5 & 6

Added a style clarifiaction for indexing
Added tradeoffs of different multidim-array styles

Grammar/consistency fixes

jrgold commented:
Fixed those up. For the multidimensional array styles, I only mentioned tradeoffs in creation/destruction/access, as small memory usage gains and performance gains seem a bit advanced at this stage.

mitchellkrome commented on arraysandpointers.tex:
```diff
@@ -1,144 +1,166 @@
+\begin{codeblock}
+int a[4] = { 0, 1, 5, 6 }; // All elements are initialized
+int b[4] = { 4, 3 }; // Remaining elements are initialized to 0
+int c[4] = { }; // All elements are initialized to 0
```
test.c:29:16: warning: ISO C forbids empty initializer braces [-Wpedantic]
     int c[4] = {};

You always have to initialise 1 element. Usually you would do something like

int c[4] = { 0 };

jrgold commented:
Whoops, forgot about that! I'll fix it up it a few hours when I have access to my computer.

joelfenwick commented:
I'll fix it when I merge.

I may release a preliminary version today, we'll see what state it's in.

---

From: jrgold notifications@github.com
Sent: Monday, 18 July 2016 8:37 AM
To: joelfenwick/ctute
Subject: Re: [joelfenwick/ctute] Refactor of chapter 6 and edits to chapter 5 (#14)

Whoops, forgot about that! I'll fix it up it a few hours when I have access to my computer.

## 

You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHubhttps://github.com/joelfenwick/ctute/pull/14#issuecomment-233208343, or mute the threadhttps://github.com/notifications/unsubscribe-auth/ARI5UAfv9cXQXXD7KO8WHoWM3DfWshpwks5qWq66gaJpZM4JOKCH.

jrgold commented:
Okay, sounds good.
